docs/design/proposals/metro-dr.md

```diff
@@ -0,0 +1,216 @@
+with the unique ID and returns the encoded value which contains all the
+required information for other operations. For metro DR, the same volume needs
+to be used in the secondary Kubernetes cluster. In the secondary cluster, the
+cephcsi cannot identify the volume corresponding to the PVC. Based on the
+assumption that the Kubernetes namespace and the PVC name remain the same when
+admin is doing the failover/fallback cephcsi can use the PVC name and namespace
+name to identify the RBD image. For that cephcsi needs to make a few changes.
+
+### Pre-req
+
+* The PVC name and namespace need to be passed to the CSI driver as part of the
+  volume create request.
+* A new storage-class parameter to indicate that image name need to be derived
+  from PVC name and namespace name.
+* An unique ID in the cephcsi ceph cluster configmap
+
+In the case of the primary cluster as the PVC name and the namespace name
+should be sent in volume create request, the cephcsi will generate omap data
+and creates its request name (PV name) and unique ID mapping in omap
+`csi.volume.<uniqueID>.PVName` : `volume UUID` (this is required to identify
+the RBD image/cephFS subvolume belongs to the PV) and creates an RBD image with
+the name `csi-vol-<pvcname>-<namespacename>-<UUID>`.  When doing the failover,
+cephcsi in the secondary cluster will check the volume is already present, If
+the image is present cephcsi will add its request name(PV Name) mapping to the
+UUID (`csi.volume.<uniqueID>.PVName` : `volume UUID`). As ID passed in
+configmap is unique across all the cephcsi instances accessing the ceph cluster
+will help cephcsi to identify the volume. Retaining the same image OMap is
+useful to preserve and reuse the volumeContext, and hence better, than creating
+a separate image OMap per peer kubernetes cluster that is sharing Ceph pool.
+
+Currently, we are making use of watchers in node stage request to make sure RWO
+PVC is mounted on a single node at a given point of time. If the unmap is not
+done on a cluster where rbd image is mounted the watcher won't be removed,
+cephcsi need to take care of handling stale watchers when the workload comes up
+on another kubernetes cluster.
+
+In case of the PVC is Retained and recreated with the same name in the same
+namespace as the request name is unique cephcsi will not return the underlying
+storage. But if the PVC is Retained on one cluster and PVC gets created with
+the same name in the same namespace in another cluster will get underlying
+storage (it won't create a new volume (the support for the need to be added
+later by marking RBD/CephFS storage for not to be reused))
+
+When cephcsi receives a request for Delete operation it will delete the backend
+storage and all OMap entries as well and thus clean up the state for all peer
+kubernetes clusters.
+
+#### Pseudo code
+
+* CreateVolume
+  * CheckVolExists
+    * if PVCNaming:
+      * List and check if the image exists, and is still unbound to a request
```

Comment:
Is this really going to list all PVs?

Reply:
Yes it has to list all the rbd images and check the image belongs to different PV

Reply:
@dillaman reason for listing the images/subvolumes is that the pvc can be retained so a user can delete the pvc and the PV and rbd image will be left, if the user creates a new pvc we should not give back the same rbd image, we need to create a new image and return it. 

The reason for not naming the rbd image with the only pvc_name+namespace_name is that retained pvc can also be done by the user just by editing the pvc spec and change the Delete policy. The admin or an operator has to rename the rbd/cephfs subvolume but the admin/operator cannot update the PV spec(immutable object) to point to the newly renamed rbd/cephfs volume and currently, we don't have any mechanism to rename the cephfs subvolume.

Reply:
We should really try to avoid such operations that are not constant time. You can achieve the same effect by naming the image under some constant naming scheme and then adding metadata to the image to tie in its unique ID.

Reply:
... this is similar to what we discussed when handling async DR cases.